model/host/host.go

```diff
@@ -1330,28 +1342,91 @@ func (h *Host) Terminate(ctx context.Context, user, reason string) error {
 	return nil
 }
 
-// SetDNSName updates the DNS name for a given host. If the dnsName is empty,
-// this will no-op and will not unset the existing DNS name.
-func (h *Host) SetDNSName(ctx context.Context, dnsName string) error {
-	if h.Host == dnsName || dnsName == "" {
+func buildEC2MetadataUpdate(hostname, zone, publicIPv4, privateIPv4, ipv6 string, launchTime time.Time, volumes []VolumeAttachment) bson.M {
+	setFields := bson.M{}
+
+	if hostname != "" {
+		setFields[DNSKey] = hostname
+	}
+	if zone != "" {
+		setFields[ZoneKey] = zone
+	}
+	if !launchTime.IsZero() {
+		setFields[StartTimeKey] = launchTime
+	}
+	if publicIPv4 != "" {
+		setFields[PublicIPv4Key] = publicIPv4
+	}
+	if privateIPv4 != "" {
+		setFields[IPv4Key] = privateIPv4
+	}
+	if ipv6 != "" {
+		setFields[IPKey] = ipv6
+	}
+	if len(volumes) > 0 {
+		setFields[VolumesKey] = volumes
+	}
+
+	return setFields
+}
+
+const numMetadataFields = 7
+
+// SetEC2Metadata updates the EC2 metadata for a given host. Only non-zero
+// fields will be set.
+func (h *Host) SetEC2Metadata(ctx context.Context, params HostMetadataOptions) error {
+	setFields := buildEC2MetadataUpdate(
+		params.Hostname,
+		params.Zone,
+		params.PublicIPv4,
+		params.PrivateIPv4,
+		params.IPv6,
+		params.LaunchTime,
+		params.Volumes,
+	)
+
+	// If there is any missing data in setFields, no-op.
+	if len(setFields) < numMetadataFields {
 		return nil
 	}
 
+	if !h.NoExpiration {
```

Comment:
Would definitely be worth a comment on why unexpirable hosts are special and what additional things need to happen which can't/won't happen here to let it be marked provisioned.